rand_distr/tests/cdf.rs

```diff
@@ -0,0 +1,196 @@
+
+    let mut max_diff: f64 = 0.;
+
+    let step_points = ecdf.step_points(); // x_i in the paper
+    for i in 1..step_points.len() {
+        let (x_i, f_i) = step_points[i];
+        let (_, f_i_1) = step_points[i - 1];
+        let cdf_i = cdf(x_i);
+        let max_1 = (cdf_i - f_i).abs();
+        let max_2 = (cdf_i - f_i_1).abs();
+
+        max_diff = max_diff.max(max_1).max(max_2);
+    }
+    max_diff
+}
+
+fn kolmogorov_smirnov_statistic_discrete(ecdf: Ecdf, cdf: impl Fn(i64) -> f64) -> f64 {
+    // We implement equation (4) from [1]
+
+    let mut max_diff: f64 = 0.;
+
+    let step_points = ecdf.step_points(); // x_i in the paper
+    for i in 1..step_points.len() {
+        let (x_i, f_i) = step_points[i];
+        let (_, f_i_1) = step_points[i - 1];
+        let max_1 = (cdf(x_i as i64) - f_i).abs();
+        let max_2 = (cdf(x_i as i64 - 1) - f_i_1).abs(); // -1 is the same as -epsilon, because we have integer support
+
+        max_diff = max_diff.max(max_1).max(max_2);
+    }
+    max_diff
+}
+
+const SAMPLE_SIZE: u64 = 1_000_000;
+
+fn critical_value() -> f64 {
+    // If the sampler is correct, we expect less than 0.001 false positives (alpha = 0.001).
+    // Passing this does not prove that the sampler is correct but is a good indication.
+    1.95 / (SAMPLE_SIZE as f64).sqrt()
+}
+
+fn sample_ecdf<T>(seed: u64, dist: impl Distribution<T>) -> Ecdf
+where
+    T: AsPrimitive<f64>,
+{
+    let mut rng = rand::rngs::SmallRng::seed_from_u64(seed);
+    let samples = (0..SAMPLE_SIZE)
+        .map(|_| dist.sample(&mut rng).as_())
+        .collect();
+    Ecdf::new(samples)
+}
+
+/// Tests a distribution against an analytical CDF.
+/// The cdf has to be continuous.
+pub fn test_continuous(seed: u64, dist: impl Distribution<f64>, cdf: impl Fn(f64) -> f64) {
+    let ecdf = sample_ecdf(seed, dist);
+    let ks_statistic = kolmogorov_smirnov_statistic_continuous(ecdf, cdf);
+
+    let critical_value = critical_value();
+
+    println!("KS statistic: {}", ks_statistic);
+    println!("Critical value: {}", critical_value);
+    assert!(ks_statistic < critical_value);
+}
+
+/// Tests a distribution over integers against an analytical CDF.
+/// The analytical CDF must not have jump points which are not integers.
+pub fn test_discrete<I: AsPrimitive<f64>>(
+    seed: u64,
+    dist: impl Distribution<I>,
+    cdf: impl Fn(i64) -> f64,
+) {
+    let ecdf = sample_ecdf(seed, dist);
+    let ks_statistic = kolmogorov_smirnov_statistic_discrete(ecdf, cdf);
+
+    // This critical value is bigger than it could be for discrete distributions, but because of large sample sizes this should not matter too much
+    let critical_value = critical_value();
+
+    println!("KS statistic: {}", ks_statistic);
+    println!("Critical value: {}", critical_value);
+    assert!(ks_statistic < critical_value);
+}
+
+fn normal_cdf(x: f64, mean: f64, std_dev: f64) -> f64 {
+    0.5 * (1.0 + ((x - mean) / (std_dev * f64::consts::SQRT_2)).erf())
```

Comment:
This could use `erfc` like statrs: https://github.com/statrs-dev/statrs/blob/master/src/distribution/normal.rs#L327

It could use the [`special::Error`](https://docs.rs/special/latest/special/trait.Error.html) trait.